### DISCUSSION TITLE
Temporary slowing down PocketBase development

### DISCUSSION THREAD

**Discussion by ganigeorgiev** [pocketbase/pocketbase#3271] [Announcements]
I'm planning for the next 2-3 months to temporary slow down the PocketBase development until I manage to finalize the ongoing work on Presentator v3.

So far I've tried to work simultaneously on both projects but I'm really terrible at multi-tasking and I'm failing to keep up with both of them and their related complementary libraries and the necessary support.

I enjoy working more on PocketBase as there is more complex and interesting work to be done, but I'm also almost 6 months behind my initial roadmap for Presentator v3 (see [presentator#183](https://github.com/presentator/presentator/issues/183)) and I feel that I'll not be able to complete it anytime soon if I continue with the same pace as I'm currently working on the project.

So for the next coulple months I've decided to temporary focus and dedicate more time finalizing Presentator v3.
From the backend side of things we already have what we need for Presentator, but I'm having some difficulties with the UI/UX and constanly switching between the 2 projects doesn't help (_especially when working on PocketBase is more tempting_).

**PocketBase development still continues, but it will be mostly on the weekends, meaning that new features will be added slightly slower.**
Bugs are always with highest priority and they will be fixed as soon as they are identified.
I'll try to provide support here and on the support mail as much as I can, but keep in mind that it could take me some time to get to your question. In the past some users marked their issues/questions as "Important/Urgent/etc." but please understand that I'm not able to provide individual and dedicated support. If I can I'll try to help, but don't expect too much.

For now the short term plan for the next couple months is:
  - Finalize Presentator v3
  - Work towards PocketBase v0.19.0/v0.20.0 to include:
    - Generalized app logging (not just request logs) and new Logs preview in the Admin UI to accommodate the change
    - Rich text editor related tasks ("Files browse" TinyMCE plugin, option to return plain text excerpt from html content, eg. with `content:excerpt` `?fields` modifier in order to speed up the records listing with large texts, etc.)
    - Some non-breaking web APIs and filter rules related tasks (_the PocketBase [roadmap](https://github.com/orgs/pocketbase/projects/2/views/1) will be updated accordingly_)

If you have a blocker issue or think that something else needs to be prioritized, feel free to let me know and I'll consider it.

**Comment by mpiorowski** [comment]
You are rocking it :) Since we first used PocketBase in our work, now almsot all of our apps depend on it :D So big thank You.

Now for the things to consider, I was wondering how hard it would be to implement OAuth login server-side...cos right now we run the client side only once, for oauth one-off realtime subscription, and then we move everything to server side via loading and saving cookie. It would be amazing to just have one pocketbase instance, server side.

You already mentioned how this can be done in https://github.com/pocketbase/pocketbase/discussions/2434, but it doesn't sounds easy :) Wonder if this can be implemented into js sdk.

One more time, thanks for great work ;)

**Reply by ganigeorgiev** [reply]
As far as I know the OAuth2 Code/PKCE authentication can't be done entirely server-side because you need to sign in first with the OAuth2 provider, aka. **at some point you need to open a url with the vendor page in the browser**.

If you don't want to initialize a client-side SDK instance, you need to implement the "Manual code exchange" flow. For an example check the second tab in https://pocketbase.io/docs/authentication/#oauth2-integration. It shows a plain JS browser implementation but the same idea apply for node or any other backend handler - you'll need basically 2 actions:
  - Somewhere that will show the "Login with ..." links.
  - Somewhere that will handle the OAuth2 redirect and will exchange the temp OAuth2 code for a token (this is when the actual PocketBase authentication happens).

The "all-in-one" OAuth2 flow is just a wrapper around it to simplify the cases where redirect/deeplink is not possible or easily done (eg. in mobile, desktop, cli applications or for non-native browser environments like Adobe XD/Figma/Photoshop plugins, etc.). The main difference between those 2 flows is that in the "all-in-one" the redirect handler is builtin and we send the result back to the client through a `@oauth2` realtime subscription. In the "Manual code exchange" flow the redirect handler is your own custom action.

Suggestions to further simplify it are welcome but I can't think of other way how to perform the OAuth2 Code/PKCE authentication.

**Reply by ausminternet** [reply]
@mpiorowski idk if it's still an actual need, but: I implemented a complete server-side oauth flow, have a look at: https://github.com/ausminternet/remix-pocketbase-starter

**Comment by TMBL-DEV** [comment]
Pocketbase is looking great! No worries that you need to slow down development. It's incredible how fast you can develop with so little time. Good luck with the Presentator v3 project.

**Comment by normanmadrid** [comment]
Great work gani! I'm always excited (too excited) when I see a major update on the releases tab of this repo!

**Comment by TigerJohn3** [comment]
Gani, best of luck with the rest of your Presentator project! Thank you for everything in PocketBase and thank you for keeping the community in the loop with your endeavors.

Do you have any plans on expanding the team working on PocketBase? That way, when you do step away to work on some other things, there are still folks devoting their full effort to PocketBase development/support. I've been making small monthly donations on OpenCollective to support you, and would be more than happy to increase this if it means getting more full time efforts from highly skilled people.

No matter what is decided, I can't thank you enough for your work with this incredible project. I grow cordyceps mushrooms and I've been using Pocketbase to create an application to track all the data we possibly can throughout the growing process. I'm really excited when v1 makes its official launch :).

**Reply by ganigeorgiev** [reply]
> Do you have any plans on expanding the team working on PocketBase? 

No, at least not at this stage.

I'm reluctant to bring other developers because I just don't feel right to demand changes and free work from random volunteers.
And even if someone for some reason agree, there is no guarantee that they will stick with it and will help with the maintenance of the project.

PocketBase doesn't have a commercial goal or monetary plans and I don't think the donations will be enough to pay for the full time of even a single dedicated developer (_the last time I checked from both OpenCollective and PayPal I think the recurring monthly amount was around ~200BGN_).

And that is OK in my opinion, considering the small scope of the project and its still tinkering phase.

**Reply by TigerJohn3** [reply]
Sounds good. Thanks Gani.

**Comment by hungcrush** [comment]
When you work on Presentator V3, you will encounter many issues. Then, you will upgrade PocketBase more. Haha, I have waited for this for a long time... 😬 

P/s: Can I see what you're doing on Presentator v3 (you will make new git, or just make new git branch) ?

**Reply by ganigeorgiev** [reply]
>  Can I see what you're doing on Presentator v3 (you will make new git, or just make new git branch) ?

The same git repo will be used - similar to the v1->v2 migration, v3 will become `master` and v2 will be moved to its own branch. I'll publish it once I'm ready with the UI and the v2->v3 migration scripts.

**Reply by hungcrush** [reply]
Thank you so much

**Comment by insidesmart** [comment]
In my limited experience (of tracking opensource projects), I will rate you at the top in terms of responsiveness to the users' queries and feedback for the project. You are an inspiration.

My best wishes for the new version of Presentator. I am sure it will be awesome.

**Comment by kjolibois** [comment]
Thank you for all of your work!

**Comment by konyilwin** [comment]
I just wanted to say a big thank you for all the hard work you put into the projects.

**Comment by m-ajd** [comment]
Thank you for your great work

**Comment by pauserratgutierrez** [comment]
You are doing an amazing job! Thank you 🥇
